hbase-server/src/main/java/org/apache/hadoop/hbase/master/AssignmentManager.java

```diff
@@ -2079,16 +2085,29 @@ private void unassign(final HRegionInfo region,
         }
 
         if (logRetries) {
-          LOG.info("Server " + server + " returned " + t + " for "
-            + region.getRegionNameAsString() + ", try=" + i
-            + " of " + this.maximumAttempts, t);
+          LOG.info("Server " + server + " returned " + t + " for " + region.getRegionNameAsString()
+              + ", try=" + i + " of " + this.maximumAttempts,
+            t);
           // Presume retry or server will expire.
         }
       }
     }
-    // Run out of attempts
-    if (state != null) {
-      regionStates.updateRegionState(region, State.FAILED_CLOSE);
+
+    long sleepTime = backoffPolicy.getBackoffTime(retryConfig,
+      getFailedAttempts(encodedName, failedCloseTracker));
+    if (failedCloseCount.incrementAndGet() <= maximumAttempts && sleepTime > 0) {
+      if (failedCloseTracker.containsKey(encodedName)) {
+        // Sleep before trying unassign if this region has failed to close before
+        scheduledThreadPoolExecutor.schedule(new DelayedUnAssignCallable(this, region, state,
```

Comment:
javadoc for the method says "Send CLOSE RPC if the server is online, otherwise, offline the region.". With this delayed callables, aren't we violating that? If we hit this block, we will still be in the CLOSING state with the thread pool retrying in the background. Curious if that causes any issues in the statemachine.

Reply:
As mentioned, this mainly deals with the case where the RS is even busy to acknowledge the request from Master and tries to have a backoff before next retry to avoid the FAILED_CLOSE state if possible. In the case where we get NotServingRegionException, we still offline the region.